File: app/src/main/java/org/wikipedia/notifications/NotificationPresenter.java
package org.wikipedia.notifications;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.net.Uri;
import android.text.TextUtils;

import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.content.ContextCompat;

import org.wikipedia.Constants;
import org.wikipedia.R;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.page.PageTitle;
import org.wikipedia.talk.TalkTopicsActivity;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.ResourceUtil;
import org.wikipedia.util.StringUtil;

public final class NotificationPresenter {
    private static final String CHANNEL_ID = "MEDIAWIKI_ECHO_CHANNEL";

    public static void showNotification(@NonNull Context context, @NonNull Notification n, @NonNull String wikiSiteName) {
        String title;
        @DrawableRes int iconResId = R.drawable.ic_speech_bubbles;
        @ColorRes int iconColor = R.color.accent50;

        NotificationCompat.Builder builder = getDefaultBuilder(context, n.id(), n.type());

        title = StringUtil.fromHtml(n.getContents() != null ? n.getContents().getHeader() : "").toString();

        if (n.getContents() != null && n.getContents().getLinks() != null
                && n.getContents().getLinks().getPrimary() != null) {
            if (Notification.CATEGORY_EDIT_USER_TALK.equals(n.category())) {
                addActionForTalkPage(context, builder, n.getContents().getLinks().getPrimary(), n);
            } else {
                addAction(context, builder, n.getContents().getLinks().getPrimary(), n);
            }
        }
        if (n.getContents() != null && n.getContents().getLinks() != null
                && n.getContents().getLinks().getSecondary() != null && n.getContents().getLinks().getSecondary().size() > 0) {
            addAction(context, builder, n.getContents().getLinks().getSecondary().get(0), n);
        }
        if (n.getContents() != null && n.getContents().getLinks() != null
                && n.getContents().getLinks().getSecondary() != null && n.getContents().getLinks().getSecondary().size() > 1) {
            addAction(context, builder, n.getContents().getLinks().getSecondary().get(1), n);
        }

        Intent activityIntent = addIntentExtras(NotificationActivity.newIntent(context), n.id(), n.type());

        String s = n.category();
        if (Notification.CATEGORY_EDIT_USER_TALK.equals(s)) {
            iconResId = R.drawable.ic_edit_user_talk;
            iconColor = R.color.accent50;
        } else if (Notification.CATEGORY_REVERTED.equals(s)) {
            iconResId = R.drawable.ic_revert;
            iconColor = R.color.base20;
        } else if (Notification.CATEGORY_EDIT_THANK.equals(s)) {
            iconResId = R.drawable.ic_user_talk;
            iconColor = R.color.green50;
        } else if (Notification.CATEGORY_THANK_YOU_EDIT.equals(s)) {
            iconResId = R.drawable.ic_edit_progressive;
            iconColor = R.color.accent50;
        } else if (s.startsWith(Notification.CATEGORY_MENTION)) {
            iconResId = R.drawable.ic_mention;
            iconColor = R.color.accent50;
        } else if (Notification.CATEGORY_LOGIN_FAIL.equals(s)) {
            iconResId = R.drawable.ic_user_avatar;
            iconColor = R.color.base0;
        }

        showNotification(context, builder, (int) n.key(), wikiSiteName, title, title, iconResId, iconColor, true, activityIntent);
    }

    public static void showMultipleUnread(@NonNull Context context, int unreadCount) {
        // When showing the multiple-unread notification, we pass the unreadCount as the "id"
        // purely for analytics purposes, to get a sense of how many unread notifications are
        // typically queued up when the user has more than two of them.
        NotificationCompat.Builder builder = getDefaultBuilder(context, unreadCount, NotificationPollBroadcastReceiver.TYPE_MULTIPLE);
        showNotification(context, builder, 0, context.getString(R.string.app_name),
                context.getString(R.string.notification_many_unread, unreadCount), context.getString(R.string.notification_many_unread, unreadCount),
                R.drawable.ic_notifications_black_24dp, R.color.accent50, true,
                addIntentExtras(NotificationActivity.newIntent(context), unreadCount, NotificationPollBroadcastReceiver.TYPE_MULTIPLE));
    }

    public static Intent addIntentExtras(@NonNull Intent intent, long id, @NonNull String type) {
        return intent.putExtra(Constants.INTENT_EXTRA_NOTIFICATION_ID, id)
                .putExtra(Constants.INTENT_EXTRA_NOTIFICATION_TYPE, type);
    }

    public static NotificationCompat.Builder getDefaultBuilder(@NonNull Context context, long id, String type) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = notificationManager.getNotificationChannel(CHANNEL_ID);
            if (notificationChannel == null) {
                int importance = NotificationManager.IMPORTANCE_HIGH;
                notificationChannel = new NotificationChannel(CHANNEL_ID,
                        context.getString(R.string.notification_echo_channel_description), importance);
                notificationChannel.setLightColor(ContextCompat.getColor(context, R.color.accent50));
                notificationChannel.enableVibration(true);
                notificationManager.createNotificationChannel(notificationChannel);
            }
        }
        return new NotificationCompat.Builder(context, CHANNEL_ID)
                .setDefaults(NotificationCompat.DEFAULT_ALL)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setAutoCancel(true)
                .setDeleteIntent(NotificationPollBroadcastReceiver.getCancelNotificationPendingIntent(context, id, type));
    }

    @SuppressWarnings("checkstyle:parameternumber")
    public static void showNotification(@NonNull Context context, @NonNull NotificationCompat.Builder builder, int id,
                                        @NonNull String title, @NonNull String text, @NonNull CharSequence longText,
                                        @DrawableRes int icon, @ColorRes int color, boolean drawIconCircle, @NonNull Intent bodyIntent) {
        builder.setContentIntent(PendingIntent.getActivity(context, 0, bodyIntent, PendingIntent.FLAG_UPDATE_CURRENT))
                .setLargeIcon(drawNotificationBitmap(context, color, icon, drawIconCircle))
                .setSmallIcon(R.drawable.ic_wikipedia_w)
                .setColor(ContextCompat.getColor(context, color))
                .setContentTitle(title)
                .setContentText(text)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(longText));

        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.notify(id, builder.build());
    }

    private static void addAction(Context context, NotificationCompat.Builder builder, Notification.Link link, Notification n) {
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0,
                addIntentExtras(new Intent(Intent.ACTION_VIEW, Uri.parse(link.getUrl())), n.id(), n.type()), 0);
        String labelStr;
        if (!TextUtils.isEmpty(link.getTooltip())) {
            labelStr = StringUtil.fromHtml(link.getTooltip()).toString();
        } else {
            labelStr = StringUtil.fromHtml(link.getLabel()).toString();
        }
        builder.addAction(0, labelStr, pendingIntent);
    }

    private static void addActionForTalkPage(Context context, NotificationCompat.Builder builder, Notification.Link link, Notification n) {
        WikiSite wiki = new WikiSite(link.getUrl());
        PageTitle title = wiki.titleForUri(Uri.parse(link.getUrl()));
        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0,
                addIntentExtras(TalkTopicsActivity.newIntent(context, title.pageTitleForTalkPage()), n.id(), n.type()), 0);
        builder.addAction(0, StringUtil.fromHtml(link.getLabel()).toString(), pendingIntent);
    }

    private static Bitmap drawNotificationBitmap(@NonNull Context context, @ColorRes int color, @DrawableRes int icon, boolean drawIconCircle) {
        final int bitmapHalfSize = DimenUtil.roundedDpToPx(20);
        final int iconHalfSize = DimenUtil.roundedDpToPx(12);
        Bitmap bmp = Bitmap.createBitmap(bitmapHalfSize * 2, bitmapHalfSize * 2, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bmp);
        Paint p = new Paint();
        p.setAntiAlias(true);
        p.setColor(ContextCompat.getColor(context, drawIconCircle ? color : android.R.color.transparent));
        canvas.drawCircle(bitmapHalfSize, bitmapHalfSize, bitmapHalfSize, p);
        Bitmap iconBmp = ResourceUtil.bitmapFromVectorDrawable(context, icon, drawIconCircle ? android.R.color.white : color);
        canvas.drawBitmap(iconBmp, null, new Rect(bitmapHalfSize - iconHalfSize, bitmapHalfSize - iconHalfSize,
                bitmapHalfSize + iconHalfSize, bitmapHalfSize + iconHalfSize), null);
        iconBmp.recycle();
        return bmp;
    }

    private NotificationPresenter() {
    }
}

File: app/src/main/java/org/wikipedia/notifications/NotificationPresenter.kt
package org.wikipedia.notifications

import android.Manifest
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.*
import android.net.Uri
import android.os.Build
import androidx.activity.result.ActivityResultLauncher
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.appcompat.view.ContextThemeWrapper
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.app.PendingIntentCompat
import androidx.core.app.RemoteInput
import androidx.core.content.ContextCompat
import androidx.core.graphics.applyCanvas
import androidx.core.graphics.createBitmap
import org.wikipedia.Constants
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.auth.AccountUtil
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.diff.ArticleEditDetailsActivity
import org.wikipedia.notifications.db.Notification
import org.wikipedia.page.PageTitle
import org.wikipedia.richtext.RichTextUtil
import org.wikipedia.settings.Prefs
import org.wikipedia.talk.TalkTopicsActivity
import org.wikipedia.theme.Theme
import org.wikipedia.util.*
import org.wikipedia.util.log.L
import java.util.Locale
import java.util.concurrent.TimeUnit

object NotificationPresenter {

    private var lastPermissionRequestTime = 0L

    fun maybeRequestPermission(context: Context, launcher: ActivityResultLauncher<String>) {
        val millisSinceLastRequest = System.currentTimeMillis() - lastPermissionRequestTime
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            !Prefs.isInitialOnboardingEnabled &&
            AccountUtil.isLoggedIn &&
            (millisSinceLastRequest > TimeUnit.HOURS.toMillis(1) || millisSinceLastRequest < 0) &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            launcher.launch(Manifest.permission.POST_NOTIFICATIONS)
            lastPermissionRequestTime = System.currentTimeMillis()
        }
    }

    fun showNotification(context: Context, n: Notification, wikiSiteName: String, lang: String) {
        val notificationCategory = NotificationCategory.find(n.category)
        var activityIntent = addIntentExtras(NotificationActivity.newIntent(context), n.id, n.type)
        val builder = getDefaultBuilder(context, n.id, n.type, notificationCategory)
        val title = RichTextUtil.stripHtml(n.contents?.header.orEmpty())
        val id = n.key().toInt()

        n.contents?.links?.let {
            it.getPrimary()?.let { primary ->
                if (NotificationCategory.EDIT_USER_TALK.id == n.category) {
                    val talkWiki = WikiSite(primary.url)
                    val talkTitle = PageTitle.titleForUri(Uri.parse(primary.url), talkWiki)
                    activityIntent = addIntentExtras(TalkTopicsActivity.newIntent(context, talkTitle, Constants.InvokeSource.NOTIFICATION), n.id, n.type)
                    addActionForTalkPage(context, builder, primary, n)
                } else {
                    addAction(context, builder, primary, n)
                }
            }
            it.secondary?.let { secondary ->
                if (secondary.isNotEmpty()) {
                    addAction(context, builder, secondary[0], n)
                }
                if (secondary.size > 1) {
                    addAction(context, builder, secondary[1], n)
                }
            }
        }

        val themedContext = ContextThemeWrapper(context, if (WikipediaApp.instance.currentTheme == Theme.LIGHT) R.style.AppTheme else WikipediaApp.instance.currentTheme.resourceId)

        showNotification(context, builder, id, n.agent?.name ?: wikiSiteName, title, title, lang,
                notificationCategory.iconResId, ResourceUtil.getThemedAttributeId(themedContext, notificationCategory.iconColor), activityIntent)
    }

    fun showMultipleUnread(context: Context, unreadCount: Int) {
        // When showing the multiple-unread notification, we pass the unreadCount as the "id"
        // purely for analytics purposes, to get a sense of how many unread notifications are
        // typically queued up when the user has more than two of them.
        val builder = getDefaultBuilder(context, unreadCount.toLong(), NotificationPollBroadcastReceiver.TYPE_MULTIPLE)
        showNotification(context, builder, 0, context.getString(R.string.app_name),
                context.getString(R.string.notification_many_unread, unreadCount), context.getString(R.string.notification_many_unread, unreadCount),
                null, R.drawable.ic_notifications_black_24dp, R.color.blue600,
                addIntentExtras(NotificationActivity.newIntent(context), unreadCount.toLong(), NotificationPollBroadcastReceiver.TYPE_MULTIPLE))
    }

    fun addIntentExtras(intent: Intent, id: Long, type: String): Intent {
        return intent.putExtra(Constants.INTENT_EXTRA_NOTIFICATION_ID, id)
                .putExtra(Constants.INTENT_EXTRA_NOTIFICATION_TYPE, type)
    }

    fun getDefaultBuilder(context: Context, id: Long, type: String?, notificationCategory: NotificationCategory = NotificationCategory.SYSTEM): NotificationCompat.Builder {
        return NotificationCompat.Builder(context, notificationCategory.id)
                .setDefaults(NotificationCompat.DEFAULT_ALL)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setAutoCancel(true)
                .setDeleteIntent(NotificationPollBroadcastReceiver.getCancelNotificationPendingIntent(context, id, type))
    }

    fun showNotification(context: Context, builder: NotificationCompat.Builder, id: Int,
                         title: String, text: String, longText: CharSequence, lang: String?,
                         @DrawableRes icon: Int, @ColorRes color: Int, bodyIntent: Intent) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            return
        }
        builder.setContentIntent(PendingIntentCompat.getActivity(context, 0, bodyIntent, PendingIntent.FLAG_UPDATE_CURRENT, false))
                .setLargeIcon(drawNotificationBitmap(context, color, icon, lang.orEmpty().uppercase(Locale.getDefault())))
                .setSmallIcon(R.drawable.ic_wikipedia_w)
                .setColor(ContextCompat.getColor(context, color))
                .setContentTitle(title)
                .setContentText(text)
                .setStyle(NotificationCompat.BigTextStyle().bigText(longText))
        NotificationManagerCompat.from(context).notify(id, builder.build())
    }

    private fun addAction(context: Context, builder: NotificationCompat.Builder, link: Notification.Link, n: Notification) {
        if (UriUtil.isDiffUrl(link.url)) {
            try {
                addActionForDiffLink(context, builder, link, n)
                return
            } catch (e: Exception) {
                L.e(e)
            }
        }
        val pendingIntent = PendingIntentCompat.getActivity(context, 0,
            addIntentExtras(Intent(Intent.ACTION_VIEW, Uri.parse(link.url)), n.id, n.type),
            PendingIntent.FLAG_UPDATE_CURRENT, false)
        val labelStr: String = if (link.tooltip.isNotEmpty()) {
            StringUtil.fromHtml(link.tooltip).toString()
        } else {
            StringUtil.fromHtml(link.label).toString()
        }
        builder.addAction(0, labelStr, pendingIntent)
    }

    private fun addActionForDiffLink(context: Context, builder: NotificationCompat.Builder, link: Notification.Link, n: Notification) {
        val uri = Uri.parse(link.url)
        val title = uri.getQueryParameter("title")
        val oldRev = uri.getQueryParameter("prev").orEmpty().toLongOrNull() ?: -1
        val newRev = uri.getQueryParameter("diff").orEmpty().toLongOrNull() ?: -1

        val pendingIntent = PendingIntentCompat.getActivity(context, 0,
            addIntentExtras(ArticleEditDetailsActivity.newIntent(context, PageTitle(title, WikiSite(link.url)), -1, oldRev, newRev), n.id, n.type),
            PendingIntent.FLAG_UPDATE_CURRENT, false)
        builder.addAction(0, StringUtil.fromHtml(link.label).toString(), pendingIntent)
    }

    private fun addActionForTalkPage(context: Context, builder: NotificationCompat.Builder, link: Notification.Link, n: Notification) {
        val wiki = WikiSite(link.url)
        val title = PageTitle.titleForUri(Uri.parse(link.url), wiki)
        val pendingIntent = PendingIntentCompat.getActivity(context, 0,
            addIntentExtras(TalkTopicsActivity.newIntent(context, title, Constants.InvokeSource.NOTIFICATION), n.id, n.type),
            PendingIntent.FLAG_UPDATE_CURRENT, false)
        builder.addAction(0, StringUtil.fromHtml(link.label).toString(), pendingIntent)
    }

    private fun addActionWithDirectReply(context: Context, builder: NotificationCompat.Builder,
                                         title: PageTitle, replyTo: String, id: Int) {
        val remoteInput = RemoteInput.Builder(NotificationPollBroadcastReceiver.RESULT_KEY_DIRECT_REPLY)
            .build()
        val resultIntent = Intent(context, NotificationPollBroadcastReceiver::class.java)
            .setAction(NotificationPollBroadcastReceiver.ACTION_DIRECT_REPLY)
            .putExtra(Constants.ARG_WIKISITE, title.wikiSite)
            .putExtra(Constants.ARG_TITLE, title)
            .putExtra(NotificationPollBroadcastReceiver.RESULT_EXTRA_REPLY_TO, replyTo)
            .putExtra(NotificationPollBroadcastReceiver.RESULT_EXTRA_ID, id)
        val resultPendingIntent = PendingIntentCompat.getBroadcast(context, 0, resultIntent, PendingIntent.FLAG_UPDATE_CURRENT, false)

        val action = NotificationCompat.Action.Builder(R.drawable.ic_reply_24, context.getString(R.string.notifications_direct_reply_action), resultPendingIntent)
            .addRemoteInput(remoteInput)
            .build()
        builder.addAction(action)
    }

    fun drawNotificationBitmap(context: Context, @ColorRes color: Int, @DrawableRes icon: Int, lang: String): Bitmap {
        val bitmapHalfSize = DimenUtil.roundedDpToPx(24f)
        val iconHalfSize = DimenUtil.roundedDpToPx(14f)
        return createBitmap(bitmapHalfSize * 2, bitmapHalfSize * 2).applyCanvas {
            val p = Paint()
            p.isAntiAlias = true
            p.color = ContextCompat.getColor(context, color)

            if (lang.isNotEmpty()) {
                p.typeface = Typeface.create("sans-serif-medium", Typeface.NORMAL)
                p.textSize = DimenUtil.dpToPx(12f)
                p.strokeWidth = DimenUtil.dpToPx(1f)

                val textBounds = Rect()
                p.getTextBounds(lang, 0, lang.length, textBounds)

                val rectPadding = DimenUtil.dpToPx(4f)
                val textLeft = bitmapHalfSize.toFloat() - (textBounds.right - textBounds.left) / 2
                val textBottom = bitmapHalfSize * 2 - rectPadding - p.strokeWidth

                drawText(lang, textLeft, textBottom, p)

                p.style = Paint.Style.STROKE
                val rBounds = RectF(textLeft + textBounds.left - rectPadding, textBottom + textBounds.top - rectPadding,
                        textLeft + textBounds.right + rectPadding, textBottom + textBounds.bottom + rectPadding)
                drawRoundRect(rBounds, rectPadding, rectPadding, p)
            }

            val iconBmp = ResourceUtil.bitmapFromVectorDrawable(context, icon, color)
            drawBitmap(iconBmp, null, Rect(bitmapHalfSize - iconHalfSize, 0,
                    bitmapHalfSize + iconHalfSize, iconHalfSize * 2), null)
            iconBmp.recycle()
        }
    }
}
